src/wrapper/strings/ffi_str.rs

```diff
@@ -1,40 +1,59 @@
 use std::{
     borrow::{Borrow, Cow, ToOwned},
-    ffi,
+    ffi::{CStr, CString},
     os::raw::c_char,
 };
 
 use cesu8::{from_java_cesu8, to_java_cesu8};
 use log::debug;
 
-use crate::wrapper::strings::ffi_str;
+#[cfg(doc)]
+use crate::wrapper::strings::JavaStr;
 
-/// Wrapper for `std::ffi::CString` that also takes care of encoding between
-/// UTF-8 and Java's Modified UTF-8. As with `CString`, this implements `Deref`
-/// to `&JNIStr`.
+/// An owned null-terminated string (like [`CString`]) encoded in Java's
+/// [modified UTF-8].
+///
+/// This type is intended for constructing Java strings from Rust code. To use
+/// it, first construct an ordinary Rust [`str`] or [`String`], then use
+/// [`JNIString::new`] to convert it to the encoding that Java expects.
+///
+/// As with `CString`, this type has a borrowed counterpart, [`JNIStr`], that
+/// it coerces to.
+///
+/// [modified UTF-8]: https://en.wikipedia.org/wiki/UTF-8#Modified_UTF-8
 pub struct JNIString {
-    internal: ffi::CString,
+    internal: CString,
 }
 
-/// Wrapper for `std::ffi::CStr` that also takes care of encoding between
-/// UTF-8 and Java's Modified UTF-8.
+/// A borrowed null-terminated string (like [`CStr`]) encoded in Java's
+/// [modified UTF-8].
+///
+/// As with `CStr` and [`str`], instances of `JNIStr` cannot be created
+/// directly, but can be borrowed from either [`JNIString`] (for constructing a
+/// Java string in Rust code) or [`JavaStr`] (which refers to an existing Java
+/// string).
+///
+/// To convert a `JNIStr` into an ordinary Rust string, use the
+/// [`to_str`][Self::to_str] method. To get a view of the modified UTF-8
+/// encoding of the `JNIStr`, use the [`as_cstr`][Self::as_cstr] method to get
+/// a `CStr`, then call its [`to_bytes`][CStr::to_bytes] method.
+///
+/// Note that, as with `CStr`, this type is **not** `repr(C)`. See [the
+/// `CStr` documentation][CStr] for an explanation of what that means. (This
+/// type is `repr(transparent)`, but it wraps around a `CStr`, not a raw
+/// pointer.)
+///
+/// [modified UTF-8]: https://en.wikipedia.org/wiki/UTF-8#Modified_UTF-8
+#[repr(transparent)] // needed because `JNIStr` gets pointer-punned from `CStr`.
```

Comment:
is this strictly true - I'm not sure it would technically be a problem if it wasn't transparent, considering that for ffi you anyway need to pull out the pointer explicitly. It seems like `transparent` repr would be more important if indeed `CStr` where considered to be ffi safe.

We aren't directly transmuting the types in a way that assumes / requires them to have the same layout, we explicitly construct by pairing up `from_ptr()` and `.as_ptr()`, so hypothetically if there were some spurious extra padding added to the `JNIStr` struct that wouldn't really be a problem?

Having it be explicitly transparent could _maybe_ be misconstrued as suggesting that it's ok to transmute the types which would be another way of unwittingly bypassing the modified-utf8 decode step?

Reply:
I realized later that we rely on casting the type references as pointers and casting those. Currently it's supposed to be an implementation detail that a `CStr` is a DST based on a `[c_char]` slice which will map to fat pointers that include the length of the string. I guess hypothetically this could be changed to just be a thin `c_char` pointer in the future and the `repr(transparent)` should ensure we maintain compatibility in that case.

Reply:
[A comment in the standard library source code](https://github.com/rust-lang/rust/blob/190f4c96116a3b59b7de4881cfec544be0246d84/library/core/src/ffi/c_str.rs#L92) says `&CStr` probably will be changed to a thin pointer in the future.